webtau-groovy/src/main/groovy/com/twosigma/webtau/cli/interactive/WebTauCliInteractive.groovy

```diff
@@ -0,0 +1,278 @@
+                ConsoleOutputs.out(Color.RED, 'enter a number or a command')
+                displayCommands(allowedCommands)
+            } else {
+                return idxOrCommand
+            }
+        }
+    }
+
+    String readLine() {
+        print new AutoResetAnsiString(Color.GREEN, 'webtau', Color.YELLOW, ' > ')
+        return inReader.readLine()
+    }
+
+    void readAndHandleIdxOrCommand(Closure idxHandler) {
+        while (true) {
+            def idxOrCommand = readIdxOrCommand()
+
+            if (idxOrCommand.command) {
+                handleCommand(idxOrCommand.command)
+                return
+            }
+
+            def done = idxHandler(idxOrCommand.idx)
+            if (done) {
+                return
+            }
+        }
+    }
+
+    private handleCommand(InteractiveCommand command) {
+        switch (command) {
+            case InteractiveCommand.Back:
+                handleBackCommand()
+                break
+            case InteractiveCommand.Quit:
+                state = State.Done
+                break
+            case InteractiveCommand.Run:
+                runSelectedScenario()
+                break
+        }
+    }
+
+    void handleBackCommand() {
+        switch (state) {
+            case State.ScenarioSelection:
+                state = State.TestSelection
+                break
+            case State.ScenarioActionSelection:
+                state = State.ScenarioSelection
+                break
+            case State.AfterRun:
+                state = State.ScenarioSelection
+                break
+        }
+    }
+
+    void displayTestFiles() {
+        ConsoleOutputs.out(Color.BLUE, 'Test files:')
+
+        testFilePaths.eachWithIndex { path, idx ->
+            ConsoleOutputs.out(Color.YELLOW, idx, Color.PURPLE, ' ', path)
+        }
+    }
+
+    List<StandaloneTest> displayScenarios(int pathIdx) {
+        def filePath = testFilePaths[pathIdx]
+        ConsoleOutputs.out(Color.BLUE, 'Test scenarios of ', Color.PURPLE, filePath, Color.BLUE, ':')
+
+        def tests = testByPath[filePath]
+        tests.eachWithIndex { test, idx ->
+            ConsoleOutputs.out(Color.YELLOW, idx, Color.PURPLE, ' ', test.scenario)
+        }
+
+        return tests
+    }
+
+    List<InteractiveCommand> commandsForCurrentState() {
+        switch (state) {
+            case State.TestSelection:
+            case State.ScenarioSelection:
+                return [InteractiveCommand.Back, InteractiveCommand.Quit]
+            case State.AfterRun:
+            case State.ScenarioActionSelection:
+                return [InteractiveCommand.Run, InteractiveCommand.Watch,
+                        InteractiveCommand.Back, InteractiveCommand.Quit]
+        }
+
+        return [InteractiveCommand.Quit]
+    }
+
+    static void displayCommands(List<InteractiveCommand> commands) {
+        commands.each this.&displayCommand
```

Comment:
Interesting that groovy allows this in a static context, confusing.

Reply:
I will remove this to avoid confusion